Library/src/Multiplayer/Components/ConversationSpaceComponent.cpp

```diff
@@ -274,4 +751,121 @@ const csp::common::String& ConversationSpaceComponent::GetConversationId() const
+						   "The Comment asset collection creation was not successful. ResCode: %d, HttpResCode: %d",
+						   static_cast<int>(AddCommentResult.GetResultCode()),
+						   AddCommentResult.GetHttpResultCode());
+
+			const MessageResult InternalResult(AddCommentResult.GetResultCode(), AddCommentResult.GetHttpResultCode());
+			INVOKE_IF_NOT_NULL(Callback, InternalResult);
+		}
+	};
+
+	const auto AssetSystem				 = csp::systems::SystemsManager::Get().GetAssetSystem();
+	const auto UniqueAssetCollectionName = ConversationSystemHelpers::GetUniqueMessageAssetCollectionName(Space.Id, UserId);
+	MessageInfo DefaultMessageInfo		 = MessageInfo();
+	DefaultMessageInfo.EditedTimestamp	 = "";
+	DefaultMessageInfo.Message			 = Message;
+	DefaultMessageInfo.UserId			 = UserId;
+	DefaultMessageInfo.IsConversation	 = false;
+	const auto MessageMetadata			 = ConversationSystemHelpers::GenerateMessageAssetCollectionMetadata(DefaultMessageInfo);
+
+	AssetSystem->CreateAssetCollection(Space.Id,
+									   ConversationId,
+									   UniqueAssetCollectionName,
+									   MessageMetadata,
+									   csp::systems::EAssetCollectionType::COMMENT,
+									   nullptr,
+									   AddCommentCallback);
+}
+
+void ConversationSpaceComponent::DeleteMessages(const csp::common::Array<csp::systems::AssetCollection>& Messages,
+												csp::systems::NullResultCallback Callback)
+{
+	const auto MessagesCount = Messages.Size();
+
+	if (MessagesCount == 0)
+	{
+		csp::systems::NullResult InternalResult(csp::systems::EResultCode::Success, (uint16_t) csp::web::EResponseCodes::ResponseNoContent);
+		INVOKE_IF_NOT_NULL(Callback, InternalResult);
+
+		return;
+	}
+
+	const auto AssetSystem = csp::systems::SystemsManager::Get().GetAssetSystem();
+
+	auto DeletionCounterDeleter = [](size_t* ptr)
+	{
+		CSP_DELETE(ptr);
+	};
+
+	std::shared_ptr<size_t> DeletionCounter(CSP_NEW size_t, DeletionCounterDeleter);
+	*DeletionCounter = 0;
+
+	for (auto idx = 0; idx < MessagesCount; ++idx)
+	{
+		const csp::systems::NullResultCallback DeleteCommentCallback = [=](const csp::systems::NullResult& DeleteCommentResult)
+		{
+			if (DeleteCommentResult.GetResultCode() == csp::systems::EResultCode::Success)
+			{
+				++*DeletionCounter;
+
+				if (*DeletionCounter == MessagesCount)
+				{
+					INVOKE_IF_NOT_NULL(Callback, DeleteCommentResult);
+				}
+			}
+			else if (DeleteCommentResult.GetResultCode() == csp::systems::EResultCode::Failed)
+			{
+
+				CSP_LOG_FORMAT(csp::systems::LogLevel::Log,
+							   "Delete asset collection for message ID %s has failed. ResCode: %d, HttpResCode: %d",
+							   Messages[idx].Id.c_str(),
+							   static_cast<int>(DeleteCommentResult.GetResultCode()),
+							   DeleteCommentResult.GetHttpResultCode());
+
+				++*DeletionCounter;
+
+				if (*DeletionCounter == MessagesCount)
+				{
+					INVOKE_IF_NOT_NULL(Callback, DeleteCommentResult);
+				}
+			}
+		};
+
+		AssetSystem->DeleteAssetCollection(Messages[idx], DeleteCommentCallback);
```

Comment:
Is there a reason we are doing it this way and not just calling the DELETE /api/v1/prototypes CHS endpoint? That would allow us to just pass in an array of the prototype Ids we want to delete and delete them all in one go.

Looking at the AssetSystem we haven't yet exposed that endpoint but it's easy enough to do.
https://ogs-odev-internal.magnoboard.com/mag-prototype/swagger/index.html

Reply:
Yes I definitely think we should take this as an opportunity to improve things if we can. Thanks.